### ISSUE TITLE
Multi source Argo CD applications fail when promotions are applied

### ISSUE THREAD

**Issue by danielloader** [akuity/kargo#2423]
# Checklist

* [x] I've searched the issue queue to verify this is not a duplicate bug report.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `kargo version`.
* [x] I've pasted logs, if applicable.

# Description

I have been attempting to use Kargo with a multi-source Argo CD application last week that has a relatively simple use-case:
- One helm chart with an upstream application (OpenFGA)
- One kustomization manifest containing auxiliary manifests for this chart e.g. Cloudnative PG cluster, External Secrets, etc.

The screenshots show the pre-promotion state (the default state of the Application) and the post promotion state of the Application. In this scenario moving from 0.2.9 to 0.2.10 on OpenFGAs helm chart, and moving from the HEAD to a fixed commit on the main branch on the kustomization.

On promotion, the targetRevisions are being bumped correctly but the rest of the object is lost causing Argo CD to fail to sync (thankfully) due to nil pointer dereferences.

```
InvalidSpecError: spec.source.repoURL and either source.path, source.chart, or source.ref are required for source {https://github.com/example/platform-services e0b568919388ed0f514bc1ac18dbed344ecbba6c nil &ApplicationSourceKustomize{NamePrefix:,NameSuffix:,Images:[],CommonLabels:map[string]string{},Version:,CommonAnnotations:map[string]string{},ForceCommonLabels:false,ForceCommonAnnotations:false,Namespace:,CommonAnnotationsEnvsubst:false,Replicas:[]KustomizeReplica{},Patches:[]KustomizePatch{},Components:[],LabelWithoutSelector:false,} nil nil }
```

# Screenshots

![image](https://github.com/user-attachments/assets/64ee585b-8387-4ff0-b2c8-3b49efdc3411)
![image](https://github.com/user-attachments/assets/dfc5854f-8a0e-4502-8191-a66c4c0678fb)

# Steps to Reproduce

Here's the stage in question, I'm assuming the warehouse configuration is not pertinent as it's finding/selecting the correct versions for the sources.

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: uat
  namespace: service-openfga
spec:
  promotionMechanisms:
    argoCDAppUpdates:
      - appName: openfga-uat
        appNamespace: argocd
        origin:
          kind: Warehouse
          name: openfga
        sourceUpdates:
          - repoURL: https://openfga.github.io/helm-charts
            chart: openfga
            updateTargetRevision: true
          - repoURL: https://github.com/example/platform-services
            updateTargetRevision: true
  requestedFreight:
    - origin:
        kind: Warehouse
        name: openfga
      sources:
        direct: true
  subscriptions: {}
```

And the relevant ignore section on the ApplicationSet:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: service-openfga
  namespace: argocd
spec:
  ignoreApplicationDifferences:
    - jqPathExpressions:
        - .spec.sources[].targetRevision
...
```

# Version

```
0.8.4
```

# Logs

```
Paste any relevant application logs here.
```

**Comment by gnadaban** [comment]
Hey @danielloader , I've #2160 open to add support for multi-source applications, take a look and 👍🏼  if you need this too. :)

**Comment by danielloader** [comment]
Thank you for saving me from the madness.

**Comment by hiddeco** [comment]
#2160 is different from the issue described here however, as that focuses on assessing the Application health.

For the issue described here, I created #2428.

Although I do have to mention a caveat remains. If the Application is managed by an ApplicationSet, and a change is made to the `sources` (for example, to add Helm values), the Application will still be reset to the revision as declared in the ApplicationSet due to an existing limitation in upstream Argo CD (https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Controlling-Resource-Modification/#limitations-of-ignoreapplicationdifferences).

**Comment by danielloader** [comment]
I'd say it's still somewhat related as the failure happened because the helm and kustomize objects became empty which itself failed the promotion.

**Comment by krancour** [comment]
> I'd say it's still somewhat related...

Just to clear up some confusion here, as @hiddeco mentioned, #2160 is focused on heath assessment of multi-source Apps. That's not related to this issue. This issue is caused by a very low-level bug in how we patch Application resources and has no bearing on how Application health is assessed.

**Comment by danielloader** [comment]
That makes sense, thanks for clearing it up and thanks for looking at it.

**Comment by hiddeco** [comment]
This precise issue should have been addressed by #2428. For the full support of multi-source Applications, we still need to put the icing on the cake of #2160.

Thanks for your report @danielloader 🙇

**Comment by danielloader** [comment]
Thanks, looking forward to the release to try it.

I'm sure I'll be back with issues and reproduction steps in future 😅
